ethcore/src/engines/clique/tests.rs

```diff
@@ -0,0 +1,776 @@
+}
+
+#[test]
+fn epoch_transition_reset_all_votes() {
+	let tester = CliqueTester::with(3, 1, vec!['A', 'B']);
+
+	let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &tester.genesis,
+											Some(tester.signers[&'C'].address()), 'A').unwrap();
+
+	let block = tester.new_block_and_import(CliqueBlockType::Empty, &block.header(), None, 'B').unwrap();
+	let block = tester.new_block_and_import(CliqueBlockType::Checkpoint, &block.header(), None, 'A').unwrap();
+
+	let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &block.header(),
+											Some(tester.signers[&'C'].address()), 'B').unwrap();
+
+	let tags = tester.into_tags(&tester.clique_signers(&block.header().hash()));
+	assert_eq!(&tags, &['A', 'B'], "Votes should have been reset after checkpoint");
+}
+
+#[test]
+fn unauthorized_signer_should_not_be_able_to_sign_block() {
+	let tester = CliqueTester::with(3, 1, vec!['A']);
+	assert!(tester.new_block_and_import(CliqueBlockType::Empty, &tester.genesis, None, 'B').is_err());
+}
+
+#[test]
+fn signer_should_not_be_able_to_sign_two_consequtive_blocks() {
+	let tester = CliqueTester::with(3, 1, vec!['A', 'B']);
+	let b = tester.new_block_and_import(CliqueBlockType::Empty, &tester.genesis, None, 'A').unwrap();
+	assert!(tester.new_block_and_import(CliqueBlockType::Empty, &b.header(), None, 'A').is_err());
+}
+
+#[test]
+fn recent_signers_should_not_reset_on_checkpoint() {
+	let tester = CliqueTester::with(3, 1, vec!['A', 'B']);
+
+	let block = tester.new_block_and_import(CliqueBlockType::Empty, &tester.genesis, None, 'A').unwrap();
+	let block = tester.new_block_and_import(CliqueBlockType::Empty, &block.header(), None, 'B').unwrap();
+	let block = tester.new_block_and_import(CliqueBlockType::Checkpoint, &block.header(), None, 'A').unwrap();
+	assert!(tester.new_block_and_import(CliqueBlockType::Empty, &block.header(), None, 'A').is_err());
+}
+
+// Not part of http://eips.ethereum.org/EIPS/eip-225
+#[test]
+fn bonus_consensus_should_keep_track_of_votes_before_latest_per_signer() {
+	let tester = CliqueTester::with(100, 1, vec!['A', 'B', 'C', 'D']);
+
+	// Add a vote for `E` signed by A
+	let vote = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &tester.genesis,
+										   Some(tester.signers[&'E'].address()), 'A').unwrap();
+	// Empty block signed by B`
+	let vote = tester.new_block_and_import(CliqueBlockType::Empty, &vote.header(), None, 'B').unwrap();
+
+	// Empty block signed by C`
+	let vote = tester.new_block_and_import(CliqueBlockType::Empty, &vote.header(), None, 'C').unwrap();
+
+	// Empty block signed by D`
+	let vote = tester.new_block_and_import(CliqueBlockType::Empty, &vote.header(), None, 'D').unwrap();
+
+	// Add a vote for `E` signed by A
+	let vote = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &vote.header(),
+										   Some(tester.signers[&'F'].address()), 'A').unwrap();
```

Comment:
The comment above has `E` while the code has `F`, which should it be?

Reply:
basically this tests so you don't implement `a stack` for each signer (as I did before)

```
Signer: A -> Vote: E
Signer: A -> Vote: F
Signer: B -> Vote: E
Signer: B -> Vote: F
Signer: C -> Vote: E
-> E should be a signer
```